allennlp_rc/dataset_readers/transformer_squad.py

```diff
@@ -0,0 +1,325 @@
+
+    def make_instances(
+        self, qid: str, question: str, answers: List[str], context: str, first_answer_offset: int
+    ) -> Iterable[Instance]:
+        answer_end = first_answer_offset + len(answers[0])
+
+        # tokenize context by spaces first, and then with the wordpiece tokenizer
+        # For RoBERTa, this produces a bug where every token is marked as beginning-of-sentence. The huggingface
+        # implementation has the same bug, so we're leaving it in.
+        tokenized_context = []
+        token_start = 0
+        for i, c in enumerate(context):
+            if c.isspace():
+                for wordpiece in self._tokenizer.tokenize(context[token_start:i]):
+                    if wordpiece.idx is None:
+                        tokenized_context.append(wordpiece)
+                    else:
+                        tokenized_context.append(
+                            wordpiece._replace(idx=wordpiece.idx + token_start)
+                        )
+                token_start = i + 1
+        for wordpiece in self._tokenizer.tokenize(context[token_start:]):
+            if wordpiece.idx is None:
+                tokenized_context.append(wordpiece)
+            else:
+                tokenized_context.append(wordpiece._replace(idx=wordpiece.idx + token_start))
+
+        token_answer_span_start = 0
+        while token_answer_span_start < len(tokenized_context) and (
+            tokenized_context[token_answer_span_start].idx is None
+            or tokenized_context[token_answer_span_start].idx < first_answer_offset
+        ):
+            token_answer_span_start += 1
+        if token_answer_span_start >= len(tokenized_context):
+            raise ValueError("Could not find answer tokens in context.")
+        if tokenized_context[token_answer_span_start].idx > first_answer_offset:
+            token_answer_span_start -= 1
+
+        token_answer_span_end = token_answer_span_start + 1
+        while token_answer_span_end < len(tokenized_context) and (
+            tokenized_context[token_answer_span_end].idx is None
+            or (
+                tokenized_context[token_answer_span_end].idx
+                + len(sanitize_wordpiece(tokenized_context[token_answer_span_end].text))
+                <= answer_end
+            )
+        ):
+            token_answer_span_end += 1
+        # If token_answer_span_end points beyond the end of the context, this just chooses all tokens up to the
+        # last one.
+        assert token_answer_span_end > token_answer_span_start
+
+        # Tokenize the question
+        tokenized_question = self._tokenizer.tokenize(question)
+        tokenized_question = tokenized_question[: self.max_query_length]
+        tokenized_question = [
+            token._replace(type_id=self.non_content_type_id, idx=None)
+            for token in tokenized_question
+        ]
+
+        # Stride over the context, making instances
+        # Sequences are <s> question </s> </s> question </s>.
```

Comment:
I don't understand this line.  You have question twice.  Is one supposed to be context?  I also don't see where this is happening in the logic below.  Ah, it's in `text_to_instance` - this comment should be moved there.  Probably also should just use [CLS] and [SEP] (noting that specific values for [CLS] and [SEP] is transformer-specific, if you care).

Reply:
I *think* why this comment is here is because of the `- 4` I was talking about above, but I could very well be wrong.

Reply:
I fixed the comment in eff17d1ce0d1854ea2bc063e6f82f73d43c2b67c. @ZhaofengWu is right, the reason it is here it to explain the `- 4`. In `text_to_instance` it is clear what the sequence is, so I didn't think a comment was necessary there.

I'm happy to write `[CLS]` and `[SEP]` instead of `<s>` and `</s>`, but it seems weird to use the special token configuration from RoBERTa, but then describe it with BERT jargon.

Reply:
Fair enough, do what you think is best.  I just suggested CLS and SEP because that's the name used in the code for those tokens, so it's more obvious what they're doing.  I think CLS and SEP are entering the standard NLP lexicon, while you have to be familiar with roberta's implementation details to know what <S> and </S> are.

Reply:
I find that convincing: b074d29b8c0074f893eff84652d46f7dbbe800c1